src/librbd/cache/rwl/Request.h

```diff
@@ -0,0 +1,200 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_LIBRBD_CACHE_RWL_REQUEST_H 
+#define CEPH_LIBRBD_CACHE_RWL_REQUEST_H 
+
+#include "include/Context.h"
+#include "librbd/cache/ImageCache.h"
+#include "librbd/cache/rwl/Types.h"
+#include "librbd/cache/rwl/LogOperation.h"
+
+namespace librbd {
+class BlockGuardCell;
+
+namespace cache {
+namespace rwl {
+
+class GuardedRequestFunctionContext;
+
+/**
+ * A request that can be deferred in a BlockGuard to sequence
+ * overlapping operations.
+ */
+template <typename T>
+class C_GuardedBlockIORequest : public Context {
+public:
+  T &rwl;
+  C_GuardedBlockIORequest(T &rwl);
+  ~C_GuardedBlockIORequest();
+  C_GuardedBlockIORequest(const C_GuardedBlockIORequest&) = delete;
+  C_GuardedBlockIORequest &operator=(const C_GuardedBlockIORequest&) = delete;
+
+  virtual const char *get_name() const = 0;
+  void set_cell(BlockGuardCell *cell);
+  BlockGuardCell *get_cell(void);
+  void release_cell();
+  
+private:
+  std::atomic<bool> m_cell_released = {false};
+  BlockGuardCell* m_cell = nullptr;
+};
+
+/**
+ * This is the custodian of the BlockGuard cell for this IO, and the
+ * state information about the progress of this IO. This object lives
+ * until the IO is persisted in all (live) log replicas.  User request
+ * may be completed from here before the IO persists.
+ */
+template <typename T>
+class C_BlockIORequest : public C_GuardedBlockIORequest<T> {
+public:
+  using C_GuardedBlockIORequest<T>::rwl;
+
+  io::Extents image_extents;
+  bufferlist bl;
+  int fadvise_flags;
+  Context *user_req; /* User write request */
+  ExtentsSummary<io::Extents> image_extents_summary;
+  bool detained = false;                /* Detained in blockguard (overlapped with a prior IO) */
```

Comment:
Any reason this is public?

Reply:
Currently, it is updated directly in `ReplicatedWriteLog<I>::release_guarded_request()`, let's improve the Encapsulation.